From 0f3d9762baf6d2d8eb0f9b49d7b1138a9f9ac8f3 Mon Sep 17 00:00:00 2001
From: Ajender Reddy Mallikarjuna <ajenderr@qti.qualcomm.com>
Date: Thu, 10 Sep 2026 12:30:00 +0530
Subject: [PATCH] gptp: add safe timestamp fallback when external source fails

Add runtime fallback to HWTSTAMP_TX_ON when external timestamp source
configuration fails via ioctl SIOCSHWTSTAMP.

Upstream-Status: Inappropriate [platform-specific bringup fix]

Signed-off-by: Ajender Reddy Mallikarjuna <ajenderr@qti.qualcomm.com>
---
 daemons/gptp/src/linux_hal_generic.cpp | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/daemons/gptp/src/linux_hal_generic.cpp b/daemons/gptp/src/linux_hal_generic.cpp
index d5b6647..88effdb 100644
--- a/daemons/gptp/src/linux_hal_generic.cpp
+++ b/daemons/gptp/src/linux_hal_generic.cpp
@@ -1,3 +1,9 @@
+/*
+ * Changes from Qualcomm Technologies, Inc. are provided under the following license:
+ *
+ * Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
+*/
 /******************************************************************************

   Copyright (c) 2009-2012, Intel Corporation
@@ -608,6 +614,13 @@ bool LinuxTimestamperGeneric::post_init( int ifindex, int sd,
     err = ioctl( sd, SIOCSHWTSTAMP, &device );
     GPTP_LOG_INFO("post_init:: SIOCSHWTSTAMP ioctl called");
 
+    if ( err == -1 && tsc_enable ) {
+        GPTP_LOG_WARNING("External timestamp source not supported (%s), fallback to HWTSTAMP_TX_ON",
+                         strerror(errno));
+        hwconfig.tx_type = HWTSTAMP_TX_ON;
+        err = ioctl( sd, SIOCSHWTSTAMP, &device );
+    }
+
     if ( err == -1 ) {
         GPTP_LOG_ERROR
         ("Failed to configure timestamping: %s", strerror(errno));
